src/astring.js

```diff
@@ -253,10 +253,13 @@ export const GENERATOR = {
     for (let i = 0; i < length; i++) {
       const statement = statements[i]
       if (writeComments && statement.comments != null) {
-        formatComments(state, statement.comments, indent, lineEnd)
+        formatComments(state, statement.comments.filter(comment => comment.end < statement.start), indent, lineEnd)
```

Comment:
IMHO, instead of filtering comments every time, it'd be more efficient to apply the logic in the filter predicate inside the `formatComments` function.